### DISCUSSION TITLE
React process.env.NODE_ENV

### DISCUSSION THREAD

**Discussion by maxim-semikov** [web-infra-dev/rsbuild#2035] [Q&A] [answered]
Hello!
Can anyone tell me how to use process.env in react application.
I created the app using the `npm create rsbuild@latest` command.
and tried to use a simple piece of code in the App.jsx file as in the documentation 
```js
  if (process.env.NODE_ENV === 'development') {
  console.log('this is a development log');
}
```

and I got an error: Uncaught ReferenceError: process is not defined

**Comment by chenjiahan** [comment]
Hi, I can not reproduce this issue, can you provide a reproduction repo?

![image](https://github.com/web-infra-dev/rsbuild/assets/7237365/7018709d-7c78-4779-8aa6-c19e2222ccef)

**Comment by maxim-semikov** [comment]
<img width="346" alt="image" src="https://github.com/web-infra-dev/rsbuild/assets/10222003/f8cf23e6-c810-4a68-a5c9-90b36e16a655">
<img width="578" alt="image" src="https://github.com/web-infra-dev/rsbuild/assets/10222003/37263797-a83d-4ccb-901f-fdbe77e8eb57">

<img width="578" alt="image" src="https://github.com/web-infra-dev/rsbuild/assets/10222003/367988fa-f80b-45cd-91c4-cf7959de3e09">



this is repo https://github.com/maxim-semikov/rsbuild-env

**Reply by xc2** [reply, marked as answer]
please use `process.env.NODE_ENV === 'development'` instead of `process?.env?.NODE_ENV === 'development'`.

This is a common misconception about the "webpack definePlugin".

nowadays the term "define" in js building field typically means replacing expression instead of declare a variable.

e.g.

`define: {"process.env.NODE_ENV": '"production"'}` will transform `process.env.NODE_ENV === 'development'` into `"production" === 'development'` but not `const process = {env: {NODE_ENV: "production"}}; process.env.NODE_ENV === 'development'`;


`process?.env?.NODE_ENV` doesn't match the expression `process.env.NODE_ENV` so it is not replaced.

**Reply by maxim-semikov** [reply]
Thanks!

**Reply by hussain-nz** [reply]
@maxim-semikov If you would like to safely access process.env (because some env variables aren't available), you can use:
> typeof process !== "undefined" && process.env.NODE_ENV
